File: examples/typescript/order-repository-dto/order/order.ts
export type LineItem = {
  sku: string;
  quantity: number;
};

export class Order {
  private constructor(
    private readonly idValue: string,
    private statusValue: string,
    private readonly customerNameValue: string,
    private readonly shippingStreetValue: string,
    private readonly itemsValue: LineItem[]
  ) {}

  static create(
    id: string,
    status: string,
    customerName: string,
    shippingStreet: string,
    items: LineItem[]
  ): Order {
    return new Order(id, status, customerName, shippingStreet, [...items]);
  }

  id(): string {
    return this.idValue;
  }

  status(): string {
    return this.statusValue;
  }

  customerName(): string {
    return this.customerNameValue;
  }

  shippingStreet(): string {
    return this.shippingStreetValue;
  }

  items(): LineItem[] {
    return [...this.itemsValue];
  }

  markReadyForFulfillment(): void {
    if (this.statusValue !== "pending") {
      throw new Error(`order must be pending before fulfillment, got ${this.statusValue}`);
    }

    this.statusValue = "ready-for-fulfillment";
  }
}

File: examples/go/order-repository-dto/order/order.go
package order

import "fmt"

type LineItem struct {
	SKU      string
	Quantity int
}

type Order struct {
	id             string
	status         string
	customerName   string
	shippingStreet string
	items          []LineItem
}

func New(id string, status string, customerName string, shippingStreet string, items []LineItem) Order {
	return Order{
		id:             id,
		status:         status,
		customerName:   customerName,
		shippingStreet: shippingStreet,
		items:          append([]LineItem(nil), items...),
	}
}

func (o Order) ID() string {
	return o.id
}

func (o Order) Status() string {
	return o.status
}

func (o Order) CustomerName() string {
	return o.customerName
}

func (o Order) ShippingStreet() string {
	return o.shippingStreet
}

func (o Order) Items() []LineItem {
	return append([]LineItem(nil), o.items...)
}

func (o *Order) MarkReadyForFulfillment() error {
	if o.status != "pending" {
		return fmt.Errorf("order must be pending before fulfillment, got %s", o.status)
	}

	o.status = "ready-for-fulfillment"
	return nil
}
